### ISSUE TITLE
Should the reference model of GRPO be an actual model?

### ISSUE THREAD

**Issue by shaojun0** [huggingface/trl#3939]
### Feature request

Add reference models to GRPO trainer for data only reference models

### Motivation

I noticed that the rel_madel design in the source code doesn't seem to be satisfactory. The original intention of referencing model design is that when the model is only used for reinforcement learning training, the initial effect is not good, so it needs to be referenced as a guarantee. The guarantee method is to use KL divergence as a support. In principle, KL divergence is equivalent to cross entropy to some extent, and cross entropy is the means used for fine-tuning most GPT models. When fine-tuning, the dataset used is specific data, not the model. Therefore, theoretically, referencing models can also use specific datasets, not the model. I am not very good at coding, so if there are any errors, please point them out.

### Your contribution

Not good at coding

**Comment by shaojun0** [comment]
I am also very poor in mathematics. If there are any misunderstandings, please point them out

**Comment by qgallouedec** [comment]
Sorry but I don't understand anything

**Comment by shaojun0** [comment]
> Sorry but I don't understand anything

My idea is to add the loss function of the model itself to the GRPO loss if ref_model is none. Is this feasible? I feel that this still does not violate the original intention of GRPO design.

**Comment by qgallouedec** [comment]
Not sure how you would do that? How do you take the reward into account?

**Comment by shaojun0** [comment]
> Not sure how you would do that? How do you take the reward into account?

<img width="1095" height="167" alt="Image" src="https://github.com/user-attachments/assets/6742fee8-9f43-4a8a-b313-ea6cbaf7defb" />
My idea is to replace the KL divergence with the loss of the model itself, whether it is feasible and the reward calculation method remains unchanged.

**Comment by shaojun0** [comment]
> Not sure how you would do that? How do you take the reward into account?

I don't know if my understanding is correct

**Comment by qgallouedec** [comment]
But the KL divergence is not a NLL right?

**Comment by shaojun0** [comment]
> But the KL divergence is not a NLL right?

I think they are equivalent,But I don't know how to explain it, my expression ability is not good, maybe it's also because my ideas are wrong

**Comment by shaojun0** [comment]
> But the KL divergence is not a NLL right?

Perhaps we can prove their equivalence mathematically, and I will try again later when I have time.

**Comment by qgallouedec** [comment]
They cannot be equivalent, for the simple reason that the KL depends on the ref model, while nll doesn't

**Comment by shaojun0** [comment]
> They cannot be equivalent, for the simple reason that the KL depends on the ref model, while nll doesn't

Is it possible that for some scenarios, referencing models may not provide good answers, so I need to use some existing historical data and utilize various reward functions to see if I can optimize for a better answer.

**Comment by shaojun0** [comment]
> They cannot be equivalent, for the simple reason that the KL depends on the ref model, while nll doesn't

I understand what you mean by non equivalence. I mean equivalence in mathematical form. Suppose the answer given by the model is exactly the same as the label in NLL. Does that mean equivalence in this situation

**Comment by shaojun0** [comment]
> They cannot be equivalent, for the simple reason that the KL depends on the ref model, while nll doesn't

Does this mean that the model can learn offline? Of course, I am a beginner, this is just my personal understanding.

**Comment by shaojun0** [comment]
> They cannot be equivalent, for the simple reason that the KL depends on the ref model, while nll doesn't

I roughly understand that the output of ref_madel is soft, but the output of NLL is hard. I'm very sorry for taking up your time. Thank you very much for taking the time to answer my question.

**Comment by qgallouedec** [comment]
> Is it possible that for some scenarios, referencing models may not provide good answers, so I need to use some existing historical data and utilize various reward functions to see if I can optimize for a better answer.

We don't optimize ref model, it's frozen

> I understand what you mean by non equivalence. I mean equivalence in mathematical form. Suppose the answer given by the model is exactly the same as the label in NLL. Does that mean equivalence in this situation

there is no label in grpo

> Does this mean that the model can learn offline? Of course, I am a beginner, this is just my personal understanding.

GRPO is online

> I roughly understand that the output of ref_madel is soft, but the output of NLL is hard.

I don't know what you mean by soft/hard

> Thank you very much for taking the time to answer my question.

No worry! I'd recommend to read the paper and more generally about RL, I think there are a couple misunderstandings on how this algorithm is working

**Comment by shaojun0** [comment]
> No worry! I'd recommend to read the paper and more generally about RL, I think there are a couple misunderstandings on how this algorithm is working

Actually, I still think the reference model for GRPO doesn't necessarily have to be a model. If a function's output is completely identical to that of the reference model, couldn't this function serve the purpose of the reference model? I still believe adding a reference function would offer much higher scalability.

**Comment by shaojun0** [comment]
> Actually, I still think the reference model for GRPO doesn't necessarily have to be a model. If a function's output is completely identical to that of the reference model, couldn't this function serve the purpose of the reference model? I still believe adding a reference function would offer much higher scalability.

In this manner, we could even embed logic within the 'reference function' to fall back on a different model whenever the output of a given model is unsatisfactory.

**Comment by albertvillanova** [comment]
Hey @shaojun0, thanks for the long back-and-forth here, and sorry it took a while to wrap up. 

To summarize where we landed: the KL term against a frozen reference model isn't interchangeable with an NLL/cross-entropy loss over a dataset — the KL depends on the ref model's distribution, and GRPO has no labels to compute an NLL against, so a "reference function" can't stand in for it without changing the algorithm itself. That said, if your goal is simply to run GRPO *without* a reference model, that's already supported: set `beta=0.0` in `GRPOConfig`, which drops the KL penalty and skips loading the ref model entirely (see https://github.com/huggingface/trl/blob/fceadc75c81bd103a6cf7f870eafddf1106b6065/trl/trainer/grpo_trainer.py#L772-L774). Since a pluggable "reference function" would change GRPO's semantics and there's no wider demand for it right now, I'll close this — but I'm happy to reopen if a concrete, paper-backed formulation comes up. Thanks for engaging so thoughtfully!
